src/journal/journald-client.c

```diff
@@ -0,0 +1,34 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "journald-client.h"
+#include "pcre2-util.h"
+
+int client_context_check_keep_log(ClientContext *c, const char *message)
```

Comment:
Given that messages can include null bytes I wonder why `message` is passed here without its length? If it's intentional I think it should probably be documented somewhere but it wouldn't be consistent with `journaltcl -g ...` (which handles binary messages just fine because it passes messages and their actual lengths to `pcre2_match` instead of relying on `strlen`).

Reply:
Updated `message` to be passed with its size. However, it seems `server_process_entry()` does not have access to the message's size. AFAICT `server_process_entry_meta()` checks the keywords in the buffer and assign the value to the correct variable, but this value is not (yet) available in `server_process_entry()`.

Reply:
> However, it seems server_process_entry() does not have access to the message's size

It doesn't and it leads to issues like https://github.com/systemd/systemd/issues/24112 among other things. I'm not sure whether it should be addressed in this PR though. Until sizes are taken into account I think it would make sense to add a test running something like
```
systemd-run --property LogFilterRegex=~ME bash -c 'printf "MESSAGE=\0GREP ME\n" | ncat --udp -U /run/systemd/journal/socket'
```
to make it clear that those filters can be bypassed by just appending null bytes at the beginning and maybe document this limitation as well.

Reply:
Agreed for the documentation. For the test, this:

```
systemd-run --property LogFilterRegex=~ME bash -c 'printf "MESSAGE=\0GREP ME\n" | ncat --udp -U /run/systemd/journal/socket'
```

won't print anything into the journal as the log will be truncated. This would highlight the wrong issue: this is not an issue with the filtering, but a limitation to native logging.

Reply:
> For the test, this ... won't print anything into the journal as the log will be truncated

Let me double-check. I think it ended up in the journal when I ran that. (If it's possible to rotate the journal by sending gibberish to the socket it's interesting as well)

Reply:
Having played with it a bit I think short-lived services can't be used to test this feature. Even without zero bytes the results are different. Sometimes "GREP ME" ends up in the journal:
```
systemd-run --property LogFilterRegex=~ME bash -c 'printf "MESSAGE=GREP ME\n" | ncat --udp -U /run/systemd/journal/socket'
...
Jul 27 18:32:04 C systemd[1]: run-r7aebd9c0bcc14e1abb41264c21ca8e5a.service: About to execute /usr/bin/bash -c "printf \"MESSAGE=GREP ME\\n\" | ncat --udp -U /run/systemd/journal/socket"
Jul 27 18:32:04 C systemd[1]: run-r7aebd9c0bcc14e1abb41264c21ca8e5a.service: Forked /usr/bin/bash as 393
Jul 27 18:32:04 C systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/run_2dr7aebd9c0bcc14e1abb41264c21ca8e5a_2eservice interface=org.>
Jul 27 18:32:04 C systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/run_2dr7aebd9c0bcc14e1abb41264c21ca8e5a_2eservice interface=org.>
Jul 27 18:32:04 C systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/run_2dr7aebd9c0bcc14e1abb41264c21ca8e5a_2eservice interface=org.freedesktop.DBus.Prop>
Jul 27 18:32:04 C systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/run_2dr7aebd9c0bcc14e1abb41264c21ca8e5a_2eservice interface=org.freedesktop.DBus.Prop>
Jul 27 18:32:04 C systemd[1]: run-r7aebd9c0bcc14e1abb41264c21ca8e5a.service: Changed dead -> running
Jul 27 18:32:04 C systemd[1]: run-r7aebd9c0bcc14e1abb41264c21ca8e5a.service: Job 2613 run-r7aebd9c0bcc14e1abb41264c21ca8e5a.service/start finished, result=done
Jul 27 18:32:04 C systemd[1]: Started run-r7aebd9c0bcc14e1abb41264c21ca8e5a.service.
Jul 27 18:32:04 C systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/run_2dr7aebd9c0bcc14e1abb41264c21ca8e5a_2eservice interface=org.>
Jul 27 18:32:04 C systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/run_2dr7aebd9c0bcc14e1abb41264c21ca8e5a_2eservice interface=org.>
Jul 27 18:32:04 C systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/run_2dr7aebd9c0bcc14e1abb41264c21ca8e5a_2eservice interface=org.freedesktop.DBus.Prop>
Jul 27 18:32:04 C systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/run_2dr7aebd9c0bcc14e1abb41264c21ca8e5a_2eservice interface=org.freedesktop.DBus.Prop>
Jul 27 18:32:04 C systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemoved cookie=1>
Jul 27 18:32:04 C systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemoved cookie=1334 reply_cookie=0 si>
Jul 27 18:32:04 C systemd[1]: Bus private-bus-connection: changing state RUNNING → CLOSING
Jul 27 18:32:04 C unknown[395]: GREP ME
```
and othertimes it doesn't
```
systemd-run --property LogFilterRegex=~ME bash -c 'printf "MESSAGE=GREP ME\n" | ncat --udp -U /run/systemd/journal/socket'
...
Jul 27 18:32:53 C systemd[1]: run-rad24fced7a7144b0a3ed3d3ff17c2b2b.service: About to execute /usr/bin/bash -c "printf \"MESSAGE=GREP ME\\n\" | ncat --udp -U /run/systemd/journal/socket"
Jul 27 18:32:53 C systemd[1]: run-rad24fced7a7144b0a3ed3d3ff17c2b2b.service: Forked /usr/bin/bash as 399
Jul 27 18:32:53 C systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/run_2drad24fced7a7144b0a3ed3d3ff17c2b2b_2eservice interface=org.>
Jul 27 18:32:53 C systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/run_2drad24fced7a7144b0a3ed3d3ff17c2b2b_2eservice interface=org.>
Jul 27 18:32:53 C systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/run_2drad24fced7a7144b0a3ed3d3ff17c2b2b_2eservice interface=org.freedesktop.DBus.Prop>
Jul 27 18:32:53 C systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/run_2drad24fced7a7144b0a3ed3d3ff17c2b2b_2eservice interface=org.freedesktop.DBus.Prop>
Jul 27 18:32:53 C systemd[1]: run-rad24fced7a7144b0a3ed3d3ff17c2b2b.service: Changed dead -> running
Jul 27 18:32:53 C systemd[1]: run-rad24fced7a7144b0a3ed3d3ff17c2b2b.service: Job 2669 run-rad24fced7a7144b0a3ed3d3ff17c2b2b.service/start finished, result=done
Jul 27 18:32:53 C systemd[1]: Started run-rad24fced7a7144b0a3ed3d3ff17c2b2b.service.
Jul 27 18:32:53 C systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/run_2drad24fced7a7144b0a3ed3d3ff17c2b2b_2eservice interface=org.>
Jul 27 18:32:53 C systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/run_2drad24fced7a7144b0a3ed3d3ff17c2b2b_2eservice interface=org.>
Jul 27 18:32:53 C systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/run_2drad24fced7a7144b0a3ed3d3ff17c2b2b_2eservice interface=org.freedesktop.DBus.Prop>
Jul 27 18:32:53 C systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/run_2drad24fced7a7144b0a3ed3d3ff17c2b2b_2eservice interface=org.freedesktop.DBus.Prop>
Jul 27 18:32:53 C systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemoved cookie=1>
Jul 27 18:32:53 C systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemoved cookie=1352 reply_cookie=0 si>
Jul 27 18:32:53 C systemd[1]: Bus private-bus-connection: changing state RUNNING → CLOSING
Jul 27 18:32:53 C systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/run_2drad24fced7a7144b0a3ed3d3ff17c2b2b_2eservice interface=org.freedesktop.DBus.Prop>
Jul 27 18:32:53 C systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/run_2drad24fced7a7144b0a3ed3d3ff17c2b2b_2eservice interface=org.freedesktop.DBus.Prop>
Jul 27 18:32:53 C systemd[1]: run-rad24fced7a7144b0a3ed3d3ff17c2b2b.service: Failed to send unit change signal for run-rad24fced7a7144b0a3ed3d3ff17c2b2b.service: Connection reset by peer
Jul 27 18:32:53 C systemd[1]: systemd-journald.service: Got notification message from PID 40 (FDSTORE=1)
Jul 27 18:32:53 C systemd[1]: systemd-journald.service: Added fd 15 (n/a) to fd store.
Jul 27 18:32:53 C systemd[1]: Bus private-bus-connection: changing state CLOSING → CLOSED
Jul 27 18:32:53 C systemd[1]: Got disconnect on private connection.
Jul 27 18:32:53 C systemd[1]: Received SIGCHLD from PID 399 (bash).
Jul 27 18:32:53 C systemd[1]: Child 399 (bash) died (code=exited, status=0/SUCCESS)
Jul 27 18:32:53 C systemd[1]: run-rad24fced7a7144b0a3ed3d3ff17c2b2b.service: Child 399 belongs to run-rad24fced7a7144b0a3ed3d3ff17c2b2b.service.
Jul 27 18:32:53 C systemd[1]: run-rad24fced7a7144b0a3ed3d3ff17c2b2b.service: Main process exited, code=exited, status=0/SUCCESS (success)
Jul 27 18:32:53 C systemd[1]: run-rad24fced7a7144b0a3ed3d3ff17c2b2b.service: Deactivated successfully.
```
it appears to be the same issue I ran into with log namespaces.

Reply:
That's definitely something that should be fixed, but it's currently out of my reach (and this PR scope). On the other hand, I expect this feature to be more relevant to long-lived services, but I could be mistaken.